src/Core/Content/Product/SalesChannel/Detail/ProductDetailRoute.php

```diff
@@ -74,7 +74,14 @@ public function load(string $productId, Request $request, SalesChannelContext $c
             );
 
             $this->dispatcher->dispatch($resolveVariantIdEvent);
-            $productId = $resolveVariantIdEvent->getResolvedVariantId() ?? $this->findBestVariant($productId, $context);
+
+            if ($resolveVariantIdEvent->getResolvedVariantId()) {
```

Comment:
with override method `findBestVariant`, here just update 

```suggestion
            $productId = $resolveVariantIdEvent->getResolvedVariantId() ?? $this->findBestVariant($productId, $context, $request->query->get('search'));
```